dbms/src/Storages/StorageBuffer.cpp

```diff
@@ -145,7 +147,34 @@ BlockInputStreams StorageBuffer::read(
         if (destination.get() == this)
             throw Exception("Destination table is myself. Read will cause infinite loop.", ErrorCodes::INFINITE_LOOP);
 
-        streams_from_dst = destination->read(column_names, query_info, context, processed_stage, max_block_size, num_streams);
+        const Block structure_of_destination_table =
+            allow_materialized ? destination->getSampleBlock() : destination->getSampleBlockNonMaterialized();
+
+        bool can_read_from_destination = true;
+        for (const String& column_name : column_names)
+        {
+           if (!structure_of_destination_table.has(column_name))
+           {
+               LOG_WARNING(log, "Destination table " << destination_database << "." << destination_table
+                   << " doesn't have column " << column_name << ". Data from destination table is skipped.");
+               can_read_from_destination = false;
+               break;
+           }
+           auto col = getColumn(column_name);
+           auto dst_col = structure_of_destination_table.getByName(column_name);
+           if (!dst_col.type->equals(*col.type))
+           {
+               LOG_WARNING(log, "Destination table " << destination_database << "." << destination_table
+                   << " have different type of column " << column_name << " ("
+                   << col.type->getName() << " != " << dst_col.type->getName()
+                   << "). Data from destination table is skipped.");
+               can_read_from_destination = false;
+               break;
+           }
+        }
+
+        if (can_read_from_destination)
+            streams_from_dst = destination->read(column_names, query_info, context, processed_stage, max_block_size, num_streams);
```

Comment:
I'm not sure about this change. When reading data from a BUFFER table we should process both the BUFFER and the destination table. But after altering the destination table we cannot simply join the data from these two sources because they have different structure.

Reply:
We can read common subset of columns from the destination table and apply ConvertingBlockInputStream.
I think, it will be more convenient than reading from buffer only.

Reply:
Using common subset can look strange, I reckon. For example:

```
CREATE TABLE test.dst (x UInt64, y UInt64) ENGINE = MergeTree ORDER BY tuple();
CREATE TABLE test.buffer (x UInt64, y UInt64) ENGINE = Buffer(test, dst, 1, 99999, 99999, 1, 1, 99999, 99999);
ALTER TABLE test.dst DROP COLUMN x, MODIFY COLUMN y String, ADD COLUMN z String;
SELECT x, y FROM test.buffer;  -- will produce only the column y
```

I mean `SELECT col1, col2 FROM table` always produced two-column output for any table. And Clickhouse throws an exception on trying to select non-existing column from a table.

Maybe it would be better to add missing column filled with default values to the blocks retrieved from the destination table. What do you think?

Reply:
Done.

I've added a new class `AddingDefaultBlockInputStream` and created a new source file `add_missing_defaults.cpp` to keep common code for `AddingDefaultBlockInputStream` and `AddingDefaultBlockOutputStream`.